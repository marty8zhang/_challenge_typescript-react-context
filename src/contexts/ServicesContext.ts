import React from 'react';
import services from '../services';

const ServicesContext = React.createContext(services);

export { ServicesContext as default };
