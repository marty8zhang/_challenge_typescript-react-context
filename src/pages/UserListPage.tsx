import React from 'react';
import User from '../models/User';
import ServicesContext from '../contexts/ServicesContext';
import { UserGatewayInterface } from '../gateways/UserGateway';
import UserNameFilter from '../components/UserNameFilter';
import UserNameFilterContext from '../contexts/UserNameFilterContext';
import UserList from '../components/UserList';

interface UserListState {
  userName: string,
  users: User[],
  filteredUsers: User[],
}

export default class UserListPage extends React.Component<any, UserListState> {
  private userNameChangeTimeoutId: ReturnType<typeof setTimeout> | null = null;

  constructor(props: Readonly<any>) {
    super(props);

    this.handleUserNameChange = this.handleUserNameChange.bind(this);

    this.state = {
      userName: '',
      users: [],
      filteredUsers: [],
    };
  }

  componentDidMount() {
    const { userGateway }: { userGateway: UserGatewayInterface } = this.context;

    userGateway.getUsers()
      .then((users: User[]) => {
        this.setState({
          users,
          filteredUsers: users,
        });
      })
      .catch((error) => {
        this.setState(() => {
          throw error;
        });
      });
  }

  handleUserNameChange(event: React.ChangeEvent<HTMLInputElement>): void {
    const { target } = event;
    const userName = target.value;

    this.setState({
      userName,
    });

    // For demonstration purposes only. E.g., how to delay the filtering if it involves external
    // API requests.
    if (this.userNameChangeTimeoutId) {
      clearTimeout(this.userNameChangeTimeoutId);
      this.userNameChangeTimeoutId = null;
    }
    this.userNameChangeTimeoutId = setTimeout(() => {
      this.setState((prevState) => {
        const { users } = prevState;
        const filteredUsers = users.filter(
          (user) => user.name.toLowerCase().indexOf(userName.trim().toLowerCase()) !== -1,
        );

        return {
          ...prevState,
          filteredUsers,
        };
      });
    }, 1000);
  }

  render() {
    const { userName, users, filteredUsers } = this.state;

    return (
      <UserNameFilterContext.Provider value={{
        userName,
        handleUserNameChange: this.handleUserNameChange,
        users,
        filteredUsers,
      }}
      >
        <UserNameFilter />
        <UserList />
      </UserNameFilterContext.Provider>
    );
  }
}

UserListPage.contextType = ServicesContext;
