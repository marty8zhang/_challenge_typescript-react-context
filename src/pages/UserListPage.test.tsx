import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import UserListPage from './UserListPage';
import ServicesContext from '../contexts/ServicesContext';

it('filters users by name', async () => {
  const mockedGetUsers = jest.fn().mockResolvedValue([
    {
      id: 1,
      name: 'Name 1',
      email: 'email1@example.com',
      address: {
        city: 'City 1',
      },
      company: {
        name: 'Company 1',
      },
    },
    {
      id: 2,
      name: 'Name 2',
      email: 'email2@example.com',
      address: {
        city: 'City 2',
      },
      company: {
        name: 'Company 2',
      },
    },
    {
      id: 3,
      name: 'Name 3',
      email: 'email3@example.com',
      address: {
        city: 'City 3',
      },
      company: {
        name: 'Company 3',
      },
    },
  ]);
  const mockedServices = {
    userGateway: {
      getUsers: mockedGetUsers,
    },
  };

  const { container } = render(
    <ServicesContext.Provider value={mockedServices}>
      <UserListPage />
    </ServicesContext.Provider>,
  );
  await screen.findByText('Name 1');

  const table = container.querySelector('table') as HTMLTableElement;
  expect(table.textContent).toContain('Name 2');
  expect(table).toMatchSnapshot();

  const input = screen.getByLabelText('Filter Users by Name:') as HTMLInputElement;
  userEvent.type(input, 'Name 3');
  expect(table.textContent).not.toContain('Name 1');
  expect(table.textContent).not.toContain('Name 2');

  userEvent.type(input, '4');
  expect(table.textContent).not.toContain('Name 1');
  expect(table.textContent).not.toContain('Name 2');
  expect(table.textContent).not.toContain('Name 3');

  userEvent.clear(input);
  expect(table.textContent).toContain('Name 1');
  expect(table.textContent).toContain('Name 2');
  expect(table.textContent).toContain('Name 3');
});
