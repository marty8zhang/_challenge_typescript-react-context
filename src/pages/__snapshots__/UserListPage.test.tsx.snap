// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters users by name 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Email
      </th>
      <th>
        City
      </th>
      <th>
        Company
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Name 1
      </td>
      <td>
        email1@example.com
      </td>
      <td>
        City 1
      </td>
      <td>
        Company 1
      </td>
    </tr>
    <tr>
      <td>
        Name 2
      </td>
      <td>
        email2@example.com
      </td>
      <td>
        City 2
      </td>
      <td>
        Company 2
      </td>
    </tr>
    <tr>
      <td>
        Name 3
      </td>
      <td>
        email3@example.com
      </td>
      <td>
        City 3
      </td>
      <td>
        Company 3
      </td>
    </tr>
  </tbody>
</table>
`;
